### PR TITLE
[SM-390] Project Access Policies

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to create server endpoints for creating, reading, updating, and deleting access policies for projects.  

## Code changes

- bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs
src/Core/SecretManagerFeatures/AccessPolicies/Interfaces/ICreateAccessPoliciesCommand.cs

Command and interface for creating access policies. 

- bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/UpdateAccessPolicyCommand.cs
src/Core/SecretManagerFeatures/AccessPolicies/Interfaces/IUpdateAccessPolicyCommand.cs

Command and interface for updating access policies.

- bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/DeleteAccessPolicyCommand.cs
src/Core/SecretManagerFeatures/AccessPolicies/Interfaces/IDeleteAccessPolicyCommand.cs

Command and interface for deleting access policies. 

- bitwarden_license/src/Commercial.Core/SecretManagerFeatures/SecretManagerCollectionExtensions.cs

Add create, update, and delete commands into dependency injection. 

- bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs
src/Core/Repositories/IAccessPolicyRepository.cs

A generic repository that handles making database updates for all entities that inherit from the `BaseAccessPolicy`. 

For now this is just implemented for policies targeting access policy types for projects: `UserProjectAccessPolicy`, `GroupProjectAccessPolicy`, and `ServiceAccountProjectAccessPolicy`. 

The idea being of implementing the rest of the access policy types in the same repository for service accounts. 

- bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/DeleteAccessPolicyCommandTests.cs
bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/UpdateAccessPolicyCommandTests.cs

Unit tests for commands.

- src/Api/Controllers/AccessPoliciesController.cs

Controller for access policy endpoints. This is hidden behind `[SecretsManager]`.

- src/Api/SecretManagerFeatures/Models/Request/AccessPoliciesCreateRequest.cs
src/Api/SecretManagerFeatures/Models/Request/AccessPolicyUpdateRequest.cs

Api request models.

- src/Api/SecretManagerFeatures/Models/Response/AccessPoliciesResponseModel.cs
src/Api/SecretManagerFeatures/Models/Response/AccessPolicyResponseModel.cs

Api response models.

- src/Core/Entities/AccessPolicy.cs

Updating core entities to enable passing grantee model references, so Api response models will contain the grantee ID and grantee name. 

- src/Infrastructure.EntityFramework/Models/AccessPolicy.cs

Update mappings for concrete access policy classes.

- src/Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs

Removing and moved to bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/ with all the other secret manager repository classes.

- src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs

With EF Core TPH we need the inherited classes in the database context.

- test/Api.IntegrationTest/Controllers/AccessPoliciesControllerTest.cs

Integration tests for the access policies controller.

- test/Api.IntegrationTest/Models/AccessPoliciesResult.cs 

A model to parse the JSON HTTP content results into from the integration tests.

- test/Api.Test/Controllers/AccessPoliciesControllerTests.cs

Unit tests for the access policies controller.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by coltonhurst — COMMENTED** [review]
@Thomas-Avery this looks really great, thanks for all the hard work on this.

@Hinton do you mind reviewing this?

**Review by Hinton — COMMENTED** [review]
Do we need access checks on all of these commands?

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                    .Include(e => e.OrganizationUser.User)
+                    .ToListAsync();
+            var groupEntities = await dbContext.GroupProjectAccessPolicy
+                    .Where(c => c.GrantedProjectId == id)
+                    .Include(e => e.Group)
+                    .ToListAsync();
+            var serviceAccountEntities = await dbContext.ServiceAccountProjectAccessPolicy
+                    .Where(c => c.GrantedProjectId == id)
+                    .Include(e => e.ServiceAccount)
+                    .ToListAsync();
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs:160]: This kind of defeats the purpose of EF TPH since we need to run 3 queries instead of 1?

**Comment by coltonhurst** [comment]
> Do we need access checks on all of these commands? It's fine to handle that in a follow up PR.

@Hinton are you referring to, for example, checking if the access policy exists (in `CreateAccessPoliciesCommand.cs`)? If so we do similar checks in the other commands for different entities (like projects). Should these checks not be at this layer?

**Comment by Hinton** [comment]
@coltonhurst no, checks to ensure the person adding the policy is allowed to do so.

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs (hunk 1 above)
Comment (Hinton) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs:160]: I think this should work. I haven't inspected the query (since the tests uses a memory one which doesn't provide a string based query) but my assumption is it would use the deliminator in the queries.

```csharp
dbContext.AccessPolicies.Where(ap =>
    ((UserProjectAccessPolicy)ap).GrantedProjectId == id ||
    ((GroupProjectAccessPolicy)ap).GrantedProjectId == id ||
    ((ServiceAccountProjectAccessPolicy)ap).GrantedProjectId == id);
```

It would then return the correct classes i.e. `ServiceAccountProjectAccessPolicy` which can be checked when cast back to the core entities.

File: bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    public CreateAccessPoliciesCommand(IAccessPolicyRepository accessPolicyRepository)
+    {
+        _accessPolicyRepository = accessPolicyRepository;
+    }
+
+    public async Task<List<BaseAccessPolicy>> CreateAsync(List<BaseAccessPolicy> accessPolicies)
+    {
+        foreach (var accessPolicy in accessPolicies)
+        {
+            if (await _accessPolicyRepository.AccessPolicyExists(accessPolicy))
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs:21]: We don't do this check on update. I.e. you could edit an access policy to match existing oen.

**Review by Hinton — COMMENTED** [review]
I started taking a better look at this PR. The way it's currently implemented we expose `AccessPolicies` to the client which is fine. However this also means we'll need to include the whole access policy objects to the client. And do logic both client side and server side to persist these IDs, and dedupe the lists.

An alternative approach would be to just send the ID (and possibly name) using something like:

```json
{
  "id": "GUID",
  "name": "Test project",
  "access": {
    "members": [],
    "groups": [],
    "serviceAccounts": []
  }
}
```

This could also be the POST/PUT which would let us abstract away the `AccessPolicy` concept from clients letting the client focus on just which entities have access instead of needing to keep track of the ID. We'll still need to do the integrity checks serverside but I don't see a way to avoid that either way.

File: bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs (hunk 2 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs:21]: The update will only modify Read, Write and RevisionDate via the `UpdateAccessPolicyCommand` so there is no risk of grantees or granted IDs changing.

**Comment by Hinton** [comment]
Reviewed the repository, I'll get around to reviewing the other files tomorrow.

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                    case GroupProjectAccessPolicy accessPolicy:
+                        policies.Add(Mapper.Map<Core.Entities.GroupProjectAccessPolicy>(accessPolicy));
+                        break;
+                    case ServiceAccountProjectAccessPolicy accessPolicy:
+                        policies.Add(Mapper.Map<Core.Entities.ServiceAccountProjectAccessPolicy>(accessPolicy));
+                        break;
+                    default:
+                        throw new ArgumentException("Unsupported access policy type");
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/192f48e4a35e3ed6728cd95e32680548be91a6e7

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var entity = await dbContext.AccessPolicies.Where(c => c.Id == baseAccessPolicy.Id).FirstOrDefaultAsync();
+            if (entity != null)
+            {
+                dbContext.AccessPolicies.Attach(entity);
+                entity.Write = baseAccessPolicy.Write;
+                entity.Read = baseAccessPolicy.Read;
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs:121]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/192f48e4a35e3ed6728cd95e32680548be91a6e7

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            }
+        }
+    }
+
+    public async Task ReplaceAsync(Core.Entities.BaseAccessPolicy baseAccessPolicy)
+    {
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+            var entity = await dbContext.AccessPolicies.Where(c => c.Id == baseAccessPolicy.Id).FirstOrDefaultAsync();
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/Repositories/AccessPolicyRepository.cs]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/192f48e4a35e3ed6728cd95e32680548be91a6e7

File: bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        data.AddRange(groupProjectAccessPolicies);
+        data.AddRange(serviceAccountProjectAccessPolicies);
+
+        await sutProvider.Sut.CreateAsync(data);
+
+        await sutProvider.GetDependency<IAccessPolicyRepository>().Received(1)
+            .CreateManyAsync(Arg.Is(AssertHelper.AssertPropertyEqual(data)));
+    }
+
+    [Theory]
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs:34]: Could we add a test where the input contains duplicates?

File: src/Api/SecretManagerFeatures/Models/Response/ProjectAccessPoliciesResponseModel.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+using Bit.Core.Models.Api;
+
+namespace Bit.Api.SecretManagerFeatures.Models.Response;
+
+public class ProjectAccessPoliciesResponseModel : ResponseModel
+{
+    public ProjectAccessPoliciesResponseModel(IEnumerable<BaseAccessPolicy>? baseAccessPolicies,
+        string obj = "projectAccessPolicies") : base(obj)
+    {
+        if (baseAccessPolicies == null) return;
```
Comment (Hinton) [inline src/Api/SecretManagerFeatures/Models/Response/ProjectAccessPoliciesResponseModel.cs:11]: Please use the full syntax.
```suggestion
        if (baseAccessPolicies == null) {
          return;
        }
```

File: test/Api.IntegrationTest/Models/AccessPoliciesResult.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
+﻿namespace Bit.Api.IntegrationTest.Models;
+
+public class AccessPoliciesResult
```
Comment (Hinton) [inline test/Api.IntegrationTest/Models/AccessPoliciesResult.cs:3]: Anything preventing us from using the API response models?

**Review by Hinton — COMMENTED** [review]
Just a few more comments otherwise LGTM.

File: test/Api.IntegrationTest/Models/AccessPoliciesResult.cs (hunk 8 above)
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/Models/AccessPoliciesResult.cs]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/d7d5318060d8b1c6bbe7cafb9f0d5a6cc93230aa

File: src/Api/SecretManagerFeatures/Models/Response/ProjectAccessPoliciesResponseModel.cs (hunk 7 above)
Comment (Thomas-Avery) [reply src/Api/SecretManagerFeatures/Models/Response/ProjectAccessPoliciesResponseModel.cs]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/1615d246d7a67fdaeae21939e786d99d43e04c3d

File: bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs (hunk 6 above)
Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs:34]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/1615d246d7a67fdaeae21939e786d99d43e04c3d

File: bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+            {
+                throw new BadRequestException("Resource already exists");
+            }
+        }
+
+        return await _accessPolicyRepository.CreateManyAsync(accessPolicies);
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommand.cs:43]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/1615d246d7a67fdaeae21939e786d99d43e04c3d

File: bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs [hunk 10]
```diff
@@ -51,4 +51,36 @@ await sutProvider.GetDependency<IAccessPolicyRepository>().Received(1)
 
         await sutProvider.GetDependency<IAccessPolicyRepository>().DidNotReceiveWithAnyArgs().CreateManyAsync(default);
     }
+
+
+    [Theory]
+    [BitAutoData]
+    public async Task CreateAsync_NotUnique_ThrowsException(
```
Comment (Hinton) [inline bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs:58]: Could we expand this with tests for each access policy? To ensure we better catch any potential regression in this.

Comment (Thomas-Avery) [reply bitwarden_license/test/Commercial.Core.Test/SecretManagerFeatures/AccessPolicies/CreateAccessPoliciesCommandTests.cs:64]: Resolved in https://github.com/bitwarden/server/pull/2507/commits/22ca6cbdc2eff07ce0c671c8639a3b4c78877067

**Review by Hinton — APPROVED** [review]
